File: packages/types/src/ai.ts
import { z } from 'zod';
import { IsoDateTimeSchema, UuidSchema } from './shared';

export const ConversationContextSchema = z.enum([
  'maintenance',
  'guest',
  'facilities',
  'community',
  'general',
]);
export type ConversationContext = z.infer<typeof ConversationContextSchema>;

export const ConversationSchema = z.object({
  id: UuidSchema,
  userId: UuidSchema.nullable(),
  context: ConversationContextSchema,
  startedAt: IsoDateTimeSchema,
  endedAt: IsoDateTimeSchema.nullable(),
  metadata: z.record(z.unknown()),
});
export type Conversation = z.infer<typeof ConversationSchema>;

export const MessageRoleSchema = z.enum(['user', 'assistant', 'system']);
export type MessageRole = z.infer<typeof MessageRoleSchema>;

export const MessageSchema = z.object({
  id: UuidSchema,
  conversationId: UuidSchema,
  role: MessageRoleSchema,
  content: z.string().nullable(),
  audioUrl: z.string().url().nullable(),
  createdAt: IsoDateTimeSchema,
});
export type Message = z.infer<typeof MessageSchema>;

export const AiSessionStatusSchema = z.enum(['active', 'ended', 'error']);
export type AiSessionStatus = z.infer<typeof AiSessionStatusSchema>;

export const AiSessionSchema = z.object({
  id: UuidSchema,
  userId: UuidSchema.nullable(),
  conversationId: UuidSchema.nullable(),
  geminiSessionId: z.string().nullable(),
  status: AiSessionStatusSchema,
  startedAt: IsoDateTimeSchema,
  endedAt: IsoDateTimeSchema.nullable(),
  errorLog: z.record(z.unknown()).nullable(),
});
export type AiSession = z.infer<typeof AiSessionSchema>;

File: apps/api/app/models/ai.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, String, Text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, UuidPk, utcnow


class Conversation(UuidPk, Base):
    __tablename__ = "conversations"

    user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True, index=True
    )
    context: Mapped[str] = mapped_column(String, nullable=False)
    started_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utcnow, nullable=False
    )
    ended_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    meta: Mapped[dict] = mapped_column("metadata", JSONB, default=dict, nullable=False)


class Message(UuidPk, Base):
    __tablename__ = "messages"

    conversation_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("conversations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    role: Mapped[str] = mapped_column(String, nullable=False)
    content: Mapped[str | None] = mapped_column(Text, nullable=True)
    audio_url: Mapped[str | None] = mapped_column(String, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utcnow, nullable=False
    )


class AiSession(UuidPk, Base):
    __tablename__ = "ai_sessions"

    user_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    conversation_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("conversations.id"), nullable=True
    )
    gemini_session_id: Mapped[str | None] = mapped_column(String, nullable=True)
    status: Mapped[str] = mapped_column(String, default="active", nullable=False)
    started_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=utcnow, nullable=False
    )
    ended_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    error_log: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
